internal/merkle/tree.go

```diff
@@ -0,0 +1,178 @@
+// (c) Cartesi and individual authors (see AUTHORS)
+// SPDX-License-Identifier: Apache-2.0 (see LICENSE)
+
+package merkle
+
+import (
+	"errors"
+	"fmt"
+	"math"
+
+	. "github.com/cartesi/rollups-node/internal/node/model"
+	"github.com/ethereum/go-ethereum/crypto"
+)
+
+// MAX_TREE_HEIGHT defines the maximum allowable height for a Merkle tree.
+const MAX_TREE_HEIGHT = 64
+
+// Tree represents a complete Merkle tree structure used to generate
+// the cryptographic proofs for the Cartesi Rollup Outputs.
+// It can hold up to 2^64 leaves, which are keccak256 hashes of 32 bytes.
+type Tree struct {
+	// The height of the tree, determining the number of levels
+	height uint
+	// A slice of slices storing the hash values for
+	// each node in the tree at each level
+	nodes [][]Hash
+	// A pointer to another Merkle tree consisting of pristine nodes. It is used
+	// as a strategy to optimize memory usage
+	pristine *pristineTree
+}
+
+// NewTree creates a new complete binary tree with the specified height.
+//
+// The height parameter specifies the number of levels in the tree. A height of 0
+// creates a tree with only a root node, while a higher number adds more levels.
+//
+// The function returns a pointer to the newly created Tree or an error
+// if the height exceeds MAX_TREE_HEIGHT.
+func NewTree(height uint) (*Tree, error) {
+	if height > MAX_TREE_HEIGHT {
+		return nil, fmt.Errorf("merkle: maximum tree height is %d", MAX_TREE_HEIGHT)
+	}
+	tree := new(Tree)
+	tree.height = height
+	tree.pristine = NewPristineTree(height)
+	tree.nodes = make([][]Hash, height+1) // the root is at height 0
+	tree.nodes[0] = make([]Hash, 1)       // sets the pristine node as the root
+	return tree, nil
+}
+
+// NewTreeFromLeaves creates a new Merkle tree with the specified height and
+// initializes it with the provided leaves.
+//
+// The height parameter specifies the number of levels in the tree.
+// The leaves parameter is a slice of Hash values to be stored as the leaf nodes
+// of the tree.
+//
+// If the number of leaves exceeds the capacity for the specified
+// height, the function returns an error.
+func NewTreeFromLeaves(height uint, leaves []Hash) (*Tree, error) {
+	if uint64(len(leaves)) > maxLeaves(height) {
+		return nil, errors.New("merkle: too many leaves")
+	}
+
+	tree, err := NewTree(height)
+	if err != nil {
+		return nil, err
+	}
+	if len(leaves) > 0 {
+		tree.nodes[height] = leaves
+	}
+	tree.calculateRootHash()
+	return tree, nil
+}
+
+// Push adds a new leaf node to the Merkle tree.
+//
+// It places the new leaf in the first available position from left to right
+// to maintain the tree's completeness.
+//
+// If the tree is already full it returns an error.
+func (t *Tree) Push(leaf Hash) error {
+	leafCount := len(t.nodes[t.height])
+	if t.height == 0 || uint64(leafCount) == maxLeaves(t.height) {
+		return errors.New("merkle: reached maximum capacity")
+	}
+
+	t.nodes[t.height] = append(t.nodes[t.height], leaf)
+	t.calculateRootHash()
```

Comment:
This is *very* inefficient. When you append a new leaf, you only need to update one node per level, the "rightmost" node. Here, you are recomputing every internal node.